USB: refactor to xmit-based API

bergzand opened the pull request:
### Contribution description

This API change refactors the usbdev API to supply buffers via the
usbdev_ep_xmit function. This changes from the usbdev_ep_ready call to allow
separate buffers per call. 

Main advantage is that the usbdev peripherals no longer have to allocate
oversized buffers for the endpoint data, potentially saving multiple KiB
of unused buffer space. These allocations are now the responsibility of
the individual USB interfaces in the firmware

An usbdev_ep_buf_t pseudotype is available and must
be used when defining buffers used for endpoints to adhere to the DMA alignment
restrictions often required with usb peripherals.

### Testing procedure

usbdev implementations:
 - [x] stm32
 - [x] nrf52
 - [x] sam0_common

Interfaces implemented:
- [x] CDC ECM
- [x] CDC ACM 
- [x] HID

(DFU only uses control requests and didn't need adaptations)

### Issues/PRs references

None

kaspar030 commented:
@bergzand please squash.

@dylad I'm about to ACK this, I tested usbus_cdc_ecm and didn't see any issues. Code looks fine to me. I suggest to merge as is?

bergzand commented:
> @bergzand please squash.

Squashed!

bergzand commented:
Rebased to resolve conflicts

dylad commented:
@bergzand Looks like Murdock isn't happy at all.

dylad commented on cpu/nrf52/periph/usbdev.c:
```diff
@@ -350,9 +326,10 @@ static int _set(usbdev_t *dev, usbopt_t opt,
+                            size_t len)
 {
-    nrfusb_t *usbdev = (nrfusb_t *)dev;
+    (void)len;
```
Unneeded ?

bergzand replied:
Removed

dylad commented on sys/usb/usbus/cdc/ecm/cdc_ecm_netdev.c:
```diff
@@ -142,7 +142,7 @@ static int _recv(netdev_t *netdev, void *buf, size_t max_len, void *info)
     }
     if (pktlen <= max_len) {
-        memcpy(buf, cdcecm->in_buf, pktlen);
+        memcpy(buf, cdcecm->data_out, pktlen);
```
in_buf -> data_out ?

bergzand replied:
Yeah, this was confusing before. I try to stick to USB terminology here. So OUT is from host to device and IN is from device to host. In this case the `data_out` is the buffer to receive data from the host, which is then copied to the receive buffer of netif.

dylad reviewed (commented):
Quick review.

bergzand commented:
rebased

bergzand commented:
I've added assertions to check the supplied buffer alignment. Should make things a bit less error prone and hopefully satisfy @maribu's align-crusade :)

bergzand commented:
And another rebase because of #17154

bergzand commented:
<details><summary>nrf52840dk with tests/cdc_ecm</summary>

```Shellsession
/home/koen/dev/RIOT-dev/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200"  
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2021-11-16 14:16:46,359 # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
ifconfig 4 add 2001:db8::2/64
2021-11-16 14:16:55,969 # ifconfig 4 add 2001:db8::2/64
2021-11-16 14:16:55,973 # success: added 2001:db8::2/64 to interface 4
> ping6 2001:db8::61b2:7a78:e34:2f87
2021-11-16 14:17:03,529 # ping6 2001:db8::61b2:7a78:e34:2f87
2021-11-16 14:17:03,540 # 12 bytes from 2001:db8::61b2:7a78:e34:2f87: icmp_seq=0 ttl=64 time=3.619 ms
2021-11-16 14:17:04,524 # 12 bytes from 2001:db8::61b2:7a78:e34:2f87: icmp_seq=1 ttl=64 time=1.564 ms
2021-11-16 14:17:05,511 # 12 bytes from 2001:db8::61b2:7a78:e34:2f87: icmp_seq=2 ttl=64 time=1.791 ms
2021-11-16 14:17:05,512 # 
2021-11-16 14:17:05,515 # --- 2001:db8::61b2:7a78:e34:2f87 PING statistics ---
2021-11-16 14:17:05,520 # 3 packets transmitted, 3 packets received, 0% packet loss
2021-11-16 14:17:05,524 # round-trip min/avg/max = 1.564/2.324/3.619 ms
```

</details>
<details><summary>nrf52840dk with tests/cdc_acm_stdio</summary>

```Shellsession
koen@zometeen ~/dev/RIOT-dev $ make -C tests/usbus_cdc_acm_stdio/ BOARD=nrf52840dk list-ttys
make: Entering directory '/home/koen/dev/RIOT-dev/tests/usbus_cdc_acm_stdio'
/sys/bus/usb/devices/1-3: SEGGER J-Link, serial: '000683806234', tty(s): ttyACM0
/sys/bus/usb/devices/1-4: RIOT-os.org USB device, serial: 'AD1B17D1BE432280', tty(s): ttyACM1
make: Leaving directory '/home/koen/dev/RIOT-dev/tests/usbus_cdc_acm_stdio'
koen@zometeen ~/dev/RIOT-dev $  make -C tests/usbus_cdc_acm_stdio/ BOARD=nrf52840dk term PORT=/dev/ttyACM1
make: Entering directory '/home/koen/dev/RIOT-dev/tests/usbus_cdc_acm_stdio'
/home/koen/dev/RIOT-dev/dist/tools/pyterm/pyterm -p "/dev/ttyACM1" -b "115200"  
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2021-11-16 14:19:04,723 # Connect to serial port /dev/ttyACM1
Welcome to pyterm!
Type '/exit' to exit.
ps
2021-11-16 14:19:11,501 # ps
2021-11-16 14:19:11,502 #       pid | name                 | state    Q | pri | stack  ( used) ( free) | base addr  | current     
2021-11-16 14:19:11,503 #         - | isr_stack            | -        - |   - |    512 (  200) (  312) | 0x20000000 | 0x200001c8
2021-11-16 14:19:11,504 #         1 | main                 | running  Q |   7 |   1536 (  880) (  656) | 0x200007ec | 0x20000a7c 
2021-11-16 14:19:11,505 #         2 | usbus                | bl anyfl _ |   1 |   1024 (  544) (  480) | 0x20000280 | 0x20000504 
2021-11-16 14:19:11,506 #           | SUM                  |            |     |   3072 ( 1624) ( 1448)
```

</details>
<details><summary>nrf52840dk with tests/usbus_hid</summary>

```Shellsession
koen@zometeen ~/dev/RIOT-dev $  make -C tests/usbus_hid/ BOARD=nrf52840dk term
make: Entering directory '/home/koen/dev/RIOT-dev/tests/usbus_hid'
/home/koen/dev/RIOT-dev/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200"  
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2021-11-16 14:20:23,436 # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
2021-11-16 14:20:59,258 # USB_HID rx_cb: Test argument 
2021-11-16 14:20:59,260 # Msg received via USB HID: 
2021-11-16 14:20:59,262 # Hello World
2021-11-16 14:20:59,262 # 
```

</details>

<details><summary>nucleo-f746zg with tests/usbus_cdc_ecm</summary>

```Shellsession
koen@zometeen ~/dev/RIOT-dev $  make -C tests/usbus_cdc_ecm/ BOARD=nucleo-f746zg term
make: Entering directory '/home/koen/dev/RIOT-dev/tests/usbus_cdc_ecm'
/home/koen/dev/RIOT-dev/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200"  
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2021-11-16 14:36:28,512 # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
ifconfig 4 add 2001:db8::2/64
2021-11-16 14:36:36,841 # ifconfig 4 add 2001:db8::2/64
2021-11-16 14:36:36,845 # success: added 2001:db8::2/64 to interface 4
> ping6 2001:db8::bac9:e4a1:f195:418a
2021-11-16 14:36:42,913 # ping6 2001:db8::bac9:e4a1:f195:418a
2021-11-16 14:36:42,923 # 12 bytes from 2001:db8::bac9:e4a1:f195:418a: icmp_seq=0 ttl=64 time=3.066 ms
2021-11-16 14:36:43,921 # 12 bytes from 2001:db8::bac9:e4a1:f195:418a: icmp_seq=1 ttl=64 time=1.193 ms
2021-11-16 14:36:44,921 # 12 bytes from 2001:db8::bac9:e4a1:f195:418a: icmp_seq=2 ttl=64 time=1.160 ms
2021-11-16 14:36:44,922 # 
2021-11-16 14:36:44,926 # --- 2001:db8::bac9:e4a1:f195:418a PING statistics ---
2021-11-16 14:36:44,931 # 3 packets transmitted, 3 packets received, 0% packet loss
2021-11-16 14:36:44,935 # round-trip min/avg/max = 1.160/1.806/3.066 ms
```

</details>

<details><summary>nucleo-f746zg with tests/usbus_cdc_acm_stdio</summary>

```Shellsession
koen@zometeen ~/dev/RIOT-dev $  make -C tests/usbus_cdc_acm_stdio/ BOARD=nucleo-f746zg list-ttys
make: Entering directory '/home/koen/dev/RIOT-dev/tests/usbus_cdc_acm_stdio'
/sys/bus/usb/devices/1-3: RIOT-os.org USB device, serial: '5B8017AD60B13119', tty(s): ttyACM1
/sys/bus/usb/devices/1-4: STMicroelectronics STM32 STLink, serial: '066BFF343334434257135447', tty(s): ttyACM0
make: Leaving directory '/home/koen/dev/RIOT-dev/tests/usbus_cdc_acm_stdio'
koen@zometeen ~/dev/RIOT-dev $  make -C tests/usbus_cdc_acm_stdio/ BOARD=nucleo-f746zg term PORT=/dev/ttyACM1
make: Entering directory '/home/koen/dev/RIOT-dev/tests/usbus_cdc_acm_stdio'
/home/koen/dev/RIOT-dev/dist/tools/pyterm/pyterm -p "/dev/ttyACM1" -b "115200"  
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2021-11-16 14:38:33,866 # Connect to serial port /dev/ttyACM1
ps
Welcome to pyterm!
Type '/exit' to exit.
2021-11-16 14:38:40,185 # main(): This is RIOT! (Version: 2022.01-devel-555-gb19a2-pr/usbdev/xmit)
ps
2021-11-16 14:38:40,185 # RIOT USB CDC ACM shell test
> ps
2021-11-16 14:38:40,187 # ps
2021-11-16 14:38:40,188 #       pid | name                 | state    Q | pri | stack  ( used) ( free) | base addr  | current     
2021-11-16 14:38:40,188 #         - | isr_stack            | -        - |   - |    512 (  168) (  344) | 0x20000000 | 0x200001c8
2021-11-16 14:38:40,189 #         1 | main                 | running  Q |   7 |   1536 (  876) (  660) | 0x200007f0 | 0x20000a84 
2021-11-16 14:38:40,190 #         2 | usbus                | bl anyfl _ |   1 |   1024 (  548) (  476) | 0x20000284 | 0x20000504 
2021-11-16 14:38:40,190 #           | SUM                  |            |     |   3072 ( 1592) ( 1480)

```

</details>
<details><summary>samr21-xpro with tests/usbus_cdc_ecm</summary>

```Shellsession
koen@zometeen ~/dev/RIOT-dev $ make -C tests/usbus_cdc_ecm/ BOARD=samr21-xpro term
make: Entering directory '/home/koen/dev/RIOT-dev/tests/usbus_cdc_ecm'
/home/koen/dev/RIOT-dev/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200"  
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2021-11-16 14:51:59,869 # Connect to serial port /dev/ttyACM0
^[[AWelcome to pyterm!
Type '/exit' to exit.
ifconfig 4 add 2001:db8::2/64
2021-11-16 14:52:02,605 # ifconfig 4 add 2001:db8::2/64
2021-11-16 14:52:02,609 # success: added 2001:db8::2/64 to interface 4
ping6 2001:db8::2084:5e15:db07:1e49
2021-11-16 14:52:10,749 # ping6 2001:db8::2084:5e15:db07:1e49
2021-11-16 14:52:10,761 # 12 bytes from 2001:db8::2084:5e15:db07:1e49: icmp_seq=0 ttl=64 time=4.664 ms
2021-11-16 14:52:11,758 # 12 bytes from 2001:db8::2084:5e15:db07:1e49: icmp_seq=1 ttl=64 time=1.637 ms
2021-11-16 14:52:12,759 # 12 bytes from 2001:db8::2084:5e15:db07:1e49: icmp_seq=2 ttl=64 time=1.555 ms
2021-11-16 14:52:12,759 # 
2021-11-16 14:52:12,764 # --- 2001:db8::2084:5e15:db07:1e49 PING statistics ---
2021-11-16 14:52:12,769 # 3 packets transmitted, 3 packets received, 0% packet loss
2021-11-16 14:52:12,773 # round-trip min/avg/max = 1.555/2.618/4.664 ms
```

</details>

<details><summary>samr21-xpro with tests/usbus_cdc_acm_stdio</summary>

```Shellsession
koen@zometeen ~/dev/RIOT-dev $  make -C tests/usbus_cdc_acm_stdio/ BOARD=samr21-xpro list-ttys
make: Entering directory '/home/koen/dev/RIOT-dev/tests/usbus_cdc_acm_stdio'
/sys/bus/usb/devices/1-3: Atmel Corp. EDBG CMSIS-DAP, serial: 'ATML2127031800008360', tty(s): ttyACM0
/sys/bus/usb/devices/1-4: RIOT-os.org USB device, serial: 'BB3072FFBF200CDA', tty(s): ttyACM1
make: Leaving directory '/home/koen/dev/RIOT-dev/tests/usbus_cdc_acm_stdio'
koen@zometeen ~/dev/RIOT-dev $ make -C tests/usbus_cdc_acm_stdio/ BOARD=samr21-xpro term PORT=/dev/ttyACM1
make: Entering directory '/home/koen/dev/RIOT-dev/tests/usbus_cdc_acm_stdio'
/home/koen/dev/RIOT-dev/dist/tools/pyterm/pyterm -p "/dev/ttyACM1" -b "115200"  
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2021-11-16 14:41:56,379 # Connect to serial port /dev/ttyACM1
ps
Welcome to pyterm!
Type '/exit' to exit.
ps
2021-11-16 14:42:02,425 # main(): This is RIOT! (Version: 2022.01-devel-555-gb19a2-pr/usbdev/xmit)
2021-11-16 14:42:02,426 # RIOT USB CDC ACM shell test
> 2021-11-16 14:42:02,427 # ps
2021-11-16 14:42:02,427 #       pid | name                 | state    Q | pri | stack  ( used) ( free) | base addr  | current     
2021-11-16 14:42:02,428 #         - | isr_stack            | -        - |   - |    512 (  188) (  324) | 0x20000000 | 0x200001c0
2021-11-16 14:42:02,429 #         1 | main                 | running  Q |   7 |   1536 (  788) (  748) | 0x200007f0 | 0x20000adc 
2021-11-16 14:42:02,430 #         2 | usbus                | bl anyfl _ |   1 |   1024 (  588) (  436) | 0x20000284 | 0x20000504 
2021-11-16 14:42:02,430 #           | SUM                  |            |     |   3072 ( 1564) ( 1508)
```
</details>

<details><summary>samr21-xpro with tests/usbus_hid</summary>

```Shellsession
koen@zometeen ~/dev/RIOT-dev $ make -C tests/usbus_hid/ BOARD=samr21-xpro term
make: Entering directory '/home/koen/dev/RIOT-dev/tests/usbus_hid'
/home/koen/dev/RIOT-dev/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200"  
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2021-11-16 14:52:50,941 # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
2021-11-16 14:53:02,961 # USB_HID rx_cb: Test argument 
2021-11-16 14:53:02,964 # Msg received via USB HID: 
2021-11-16 14:53:02,965 # Hello world
2021-11-16 14:53:02,965 # 
```

</details>

dylad reviewed (commented):
One remark.
Also did you test usbdev_mock ?

bergzand commented on cpu/nrf52/periph/usbdev.c:
```diff
@@ -156,20 +145,10 @@ static void _copy_setup(usbdev_ep_t *ep)
     if (setup->request == USB_SETUP_REQ_SET_ADDRESS) {
         DEBUG("nrfusb: set address call\n");
         usbdev->sstate = NRFUSB_SETUP_READY;
+#include "architecture.h"
```
> Why is this here ?

No idea, removed.

bergzand commented:
>  Also did you test usbdev_mock ?

Here you go:

```Shellsession
koen@zometeen ~/dev/RIOT-dev $ make -C tests/usbus term
make: Entering directory '/home/koen/dev/RIOT-dev/tests/usbus'
/home/koen/dev/RIOT-dev/tests/usbus/bin/native/tests_usbus.elf /dev/ttyACM0 
RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

Help: Press s to start test, r to print it is ready
s
START
main(): This is RIOT! (Version: 2022.01-devel-556-ge6d56c-pr/usbdev/xmit)
.
OK (1 tests)
```

dylad commented:
I am happy with the changes,
@bergzand provided all the tests outputs, I also run a few tests on my side.
Please squash !

dylad commented:
Or wait for #17203 to be merged to squash+rebase ?

bergzand commented:
Squashed!

> Or wait for #17203 to be merged to squash+rebase ?

I don't expect that they conflict.

dylad commented:
Then here we go !
